File: utils/rate_limiter.py
import asyncio
import time

class RateLimiter:
    def __init__(self):
        self.lock = asyncio.Lock()
        self.retry_after = 0

    async def wait(self):
        async with self.lock:
            if self.retry_after > time.time():
                await asyncio.sleep(self.retry_after - time.time())

    def update_rate_limit(self, retry_after):
        self.retry_after = time.time() + retry_after

File: utils/rate_limiter.ts
class RateLimiter {
    private lock: Promise<void> | null;
    private retryAfter: number;

    constructor() {
        this.lock = null;
        this.retryAfter = 0;
    }

    async wait(): Promise<void> {
        if (this.lock) {
            await this.lock;
        }
        if (this.retryAfter > Date.now()) {
            await new Promise(resolve => setTimeout(resolve, this.retryAfter - Date.now()));
        }
    }

    updateRateLimit(retryAfter: number): void {
        this.retryAfter = Date.now() + retryAfter;
    }
}

export default RateLimiter;
